core/src/main/java/com/linecorp/armeria/internal/common/DefaultSplitHttpResponse.java

```diff
@@ -0,0 +1,335 @@
+import com.linecorp.armeria.common.stream.StreamMessage;
+import com.linecorp.armeria.common.stream.SubscriptionOption;
+import com.linecorp.armeria.common.util.UnmodifiableFuture;
+import com.linecorp.armeria.internal.common.stream.NoopSubscription;
+
+import io.netty.util.concurrent.EventExecutor;
+
+public class DefaultSplitHttpResponse implements StreamMessage<HttpData>, SplitHttpResponse {
+
+    private static final Logger logger = LoggerFactory.getLogger(DefaultSplitHttpResponse.class);
+
+    @SuppressWarnings("rawtypes")
+    private static final AtomicReferenceFieldUpdater<BodySubscriber, Subscriber> downstreamUpdater =
+            AtomicReferenceFieldUpdater.newUpdater(BodySubscriber.class, Subscriber.class, "downstream");
+
+    @SuppressWarnings("rawtypes")
+    private static final AtomicReferenceFieldUpdater<DefaultSplitHttpResponse, HeadersFuture>
+            trailersFutureUpdater = AtomicReferenceFieldUpdater
+            .newUpdater(DefaultSplitHttpResponse.class, HeadersFuture.class, "trailersFuture");
+
+    private static final ResponseHeaders HEADERS_WITH_UNKNOWN_STATUS = ResponseHeaders.of(HttpStatus.UNKNOWN);
+    private static final HeadersFuture<HttpHeaders> EMPTY_TRAILERS;
+
+    static {
+        EMPTY_TRAILERS = new HeadersFuture<>();
+        EMPTY_TRAILERS.doComplete(HttpHeaders.of());
+    }
+
+    private final HeadersFuture<ResponseHeaders> headersFuture = new HeadersFuture<>();
+    private final BodySubscriber bodySubscriber = new BodySubscriber();
+    private final HttpResponse response;
+    private final EventExecutor executor;
+
+    @Nullable
+    private volatile HeadersFuture<HttpHeaders> trailersFuture;
+    private volatile boolean wroteAny;
+
+    public DefaultSplitHttpResponse(HttpResponse response, EventExecutor executor,
+                                    SubscriptionOption... options) {
+        this.response = requireNonNull(response, "response");
+        this.executor = requireNonNull(executor, "executor");
+
+        response.subscribe(bodySubscriber, executor, options);
+    }
+
+    @Override
+    public final CompletableFuture<ResponseHeaders> headers() {
+        return headersFuture;
+    }
+
+    @Override
+    public final StreamMessage<HttpData> body() {
+        return this;
+    }
+
+    @Override
+    public final CompletableFuture<HttpHeaders> trailers() {
+        HeadersFuture<HttpHeaders> trailersFuture = this.trailersFuture;
+        if (trailersFuture != null) {
+            return trailersFuture;
+        }
+
+        trailersFuture = new HeadersFuture<>();
+        if (trailersFutureUpdater.compareAndSet(this, null, trailersFuture)) {
+            return trailersFuture;
+        } else {
+            return this.trailersFuture;
+        }
+    }
+
+    @Override
+    public boolean isOpen() {
+        return response.isOpen();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return !isOpen() && !wroteAny;
```

Comment:
Can't this be just `response.isEmpty()`?

Reply:
I wrote this code in August. Sorry, I don't remember why I added its own `wroteAny`. 😂
Let me take some time to remind it. 😎

Reply:
Now, I got it.

`DefaultSplitHttpResponse` publishes only `HttpData`. `wroteAny` is set to true when a `HttpData` is receivied from upstream.
AFAIK, `wroteAny` in `DefaultStreamMessage` which is super class of `DefaultHttpResponse` is set to true a headers is written.

Reply:
Thanks for the explanation. 🙇